### ISSUE TITLE
Using await on an internal future makes private type trigger missing_debug_implementations lint

### ISSUE THREAD

**Issue by seanmonstar** [rust-lang/rust#54239]
```rust
#![feature(async_await, await_macro, futures_api)]
#![deny(missing_debug_implementations)]

struct DontLookAtMe(i32);

async fn secret() -> DontLookAtMe {
    DontLookAtMe(41)
}

// Comment this function out to fix the lint...
pub async fn looking() -> i32 {
    await!(secret()).0
}

fn main() {
}
```

Playground: https://play.rust-lang.org/?gist=0baad2fb680123d124806d6269c15c43&version=nightly&mode=debug&edition=2018

The `DontLookAtMe` struct isn't public, so it normally shouldn't trigger the `missing_debug_implementations` lint. But once `await!` is used on a future that happens to be holding the field somewhere inside, it trips the lint:

```
error: type does not implement `fmt::Debug`; consider adding #[derive(Debug)] or a manual implementation
 --> src/main.rs:4:1
  |
4 | struct DontLookAtMe(i32);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: lint level defined here
 --> src/main.rs:2:9
  |
2 | #![deny(missing_debug_implementations)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

**Comment by Nemo157** [comment]
[Appears to be a generators bug](https://play.rust-lang.org/?gist=4c4c908d246aa3a27afc771831fd36d7&version=nightly&mode=debug&edition=2018), not specific to `async`/`await`:

```rust
#![feature(generators, generator_trait)]
#![deny(missing_debug_implementations)]

use std::ops::Generator;

struct DontLookAtMe(i32);

fn secret() -> DontLookAtMe {
    DontLookAtMe(41)
}

// Comment this function out to fix the lint...
pub fn looking() -> impl Generator<Yield = (), Return = i32> {
    || {
        let d = secret();
        yield;
        d.0
    }
}
```

**Comment by jonas-schievink** [comment]
Updated async/await code:

```rust
#![deny(missing_debug_implementations)]

struct DontLookAtMe(i32);

async fn secret() -> DontLookAtMe {
    DontLookAtMe(41)
}

// Comment this function out to fix the lint...
pub async fn looking() -> i32 {
    secret().await.0
}

fn main() {
}
```

Neither this nor the generator example trigger the lint anymore, so this looks fixed.
